File: trunk/src/main/java/net/esper/event/property/PropertyListBuilderExplicit.java
package net.esper.event.property;

import net.esper.event.EventPropertyDescriptor;
import net.esper.event.EventPropertyType;
import net.esper.client.ConfigurationEventTypeLegacy;
import net.esper.client.ConfigurationException;

import java.util.List;
import java.util.LinkedList;
import java.util.Map;
import java.util.Arrays;
import java.lang.reflect.Field;
import java.lang.reflect.Method;

public class PropertyListBuilderExplicit implements PropertyListBuilder
{
    private ConfigurationEventTypeLegacy legacyConfig;

    public PropertyListBuilderExplicit(ConfigurationEventTypeLegacy legacyConfig)
    {
        if (legacyConfig == null)
        {
            throw new IllegalArgumentException("Required configuration not passed");
        }
        this.legacyConfig = legacyConfig;
    }

    public List<EventPropertyDescriptor> assessProperties(Class clazz)
    {
        List<EventPropertyDescriptor> result = new LinkedList<EventPropertyDescriptor>();
        getExplicitProperties(result, clazz, legacyConfig);
        return result;
    }

    protected static void getExplicitProperties(List<EventPropertyDescriptor> result,
                                                                         Class clazz,
                                                                         ConfigurationEventTypeLegacy legacyConfig)
    {
        for (ConfigurationEventTypeLegacy.LegacyFieldPropDesc desc : legacyConfig.getFieldProperties())
        {
            result.add(makeDesc(clazz, desc));
        }
        for (ConfigurationEventTypeLegacy.LegacyMethodPropDesc desc : legacyConfig.getMethodProperties())
        {
            result.add(makeDesc(clazz, desc));
        }
    }

    private static EventPropertyDescriptor makeDesc(Class clazz, ConfigurationEventTypeLegacy.LegacyMethodPropDesc methodDesc)
    {
        Method[] methods = clazz.getMethods();
        Method method = null;
        for (int i = 0; i < methods.length; i++)
        {
            if (!methods[i].getName().equals(methodDesc.getAccessorMethodName()))
            {
                continue;
            }
            if (methods[i].getReturnType() == void.class)
            {
                continue;
            }
            if (methods[i].getParameterTypes().length >= 2)
            {
                continue;
            }
            if (methods[i].getParameterTypes().length == 0)
            {
                method = methods[i];
                break;
            }

            Class parameterType = methods[i].getParameterTypes()[0];
            if ((parameterType != int.class) && ((parameterType != Integer.class)) &&
                (parameterType != String.class))
            {
                continue;
            }

            method = methods[i];
            break;
        }

        if (method == null)
        {
            throw new ConfigurationException("Configured method named '" +
                    methodDesc.getAccessorMethodName() + "' not found for class " + clazz.getName());
        }

        return makeMethodDesc(method, methodDesc.getName());
    }

    private static EventPropertyDescriptor makeDesc(Class clazz, ConfigurationEventTypeLegacy.LegacyFieldPropDesc fieldDesc)
    {
        Field field = null;
        try
        {
            field = clazz.getField(fieldDesc.getAccessorFieldName());
        }
        catch (NoSuchFieldException ex)
        {
            throw new ConfigurationException("Configured field named '" +
                    fieldDesc.getAccessorFieldName() + "' not found for class " + clazz.getName());
        }
        return makeFieldDesc(field, fieldDesc.getName());
    }

    protected static EventPropertyDescriptor makeFieldDesc(Field field, String name)
    {
        return new EventPropertyDescriptor(name, name, field, EventPropertyType.SIMPLE);
    }

    protected static EventPropertyDescriptor makeMethodDesc(Method method, String name)
    {
        Class type = method.getReturnType();
        EventPropertyType propertyType = null;

        if (method.getParameterTypes().length == 1)
        {
            Class parameterType = method.getParameterTypes()[0];
            if (parameterType == String.class)
            {
                propertyType = EventPropertyType.MAPPED;
            }
            else
            {
                propertyType = EventPropertyType.INDEXED;
            }
        }
        else
        {
            propertyType = EventPropertyType.SIMPLE;
        }

        return new EventPropertyDescriptor(name, name, method, propertyType);
    }
}

File: NEsper/events/property/PropertyListBuilderExplicit.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Reflection;

using net.esper.compat;
using net.esper.client;
using net.esper.events;

namespace net.esper.events.property
{
    /// <summary>
    /// Introspector that considers explicitly configured event properties only.
    /// </summary>

    public class PropertyListBuilderExplicit : PropertyListBuilder
    {
        private ConfigurationEventTypeLegacy legacyConfig;

        /// <summary> Ctor.</summary>
        /// <param name="legacyConfig">is a legacy type specification containing
        /// information about explicitly configured fields and methods
        /// </param>

        public PropertyListBuilderExplicit(ConfigurationEventTypeLegacy legacyConfig)
        {
            if (legacyConfig == null)
            {
                throw new ArgumentException("Required configuration not passed");
            }

            this.legacyConfig = legacyConfig;
        }

		/// <summary>
		/// Introspect the type and deterime exposed event properties.
		/// </summary>
		/// <param name="type">type to introspect</param>
		/// <returns>list of event property descriptors</returns>
        
        public IList<EventPropertyDescriptor> assessProperties(Type type)
        {
            IList<EventPropertyDescriptor> result = new List<EventPropertyDescriptor>();
            getExplicitProperties(result, type, legacyConfig);
            return result;
        }

        /// <summary>
        /// Populates explicitly-defined properties into the result list.
        /// </summary>
        /// <param name="result">is the resulting list of event property descriptors</param>
        /// <param name="type">is the class to introspect</param>
        /// <param name="legacyConfig">supplies specification of explicit methods and fields to expose</param>
        
        internal static void getExplicitProperties(
            IList<EventPropertyDescriptor> result,
            Type type,
            ConfigurationEventTypeLegacy legacyConfig )
        {
            foreach (ConfigurationEventTypeLegacy.LegacyFieldPropDesc desc in legacyConfig.FieldProperties)
            {
                result.Add(makeDesc(type, desc));
            }
            
            foreach (ConfigurationEventTypeLegacy.LegacyMethodPropDesc desc in legacyConfig.MethodProperties)
            {
                result.Add(makeDesc(type, desc));
            }
        }

        private static EventPropertyDescriptor makeDesc(Type type, ConfigurationEventTypeLegacy.LegacyMethodPropDesc methodDesc)
        {
            MethodInfo[] methods = type.GetMethods();
            MethodInfo method = null;

            foreach( MethodInfo _mi in methods )
            {
                if (!_mi.Name.Equals(methodDesc.AccessorMethodName))
                {
                    continue;
                }
                if (_mi.ReturnType == typeof(void))
                {
                    continue;
                }
                if (_mi.GetParameters().Length >= 2)
                {
                    continue;
                }
                if (_mi.GetParameters().Length == 0)
                {
                    method = _mi;
                    break;
                }

                Type parameterType = _mi.GetParameters()[0].ParameterType;
                if ((parameterType != typeof(int)) &&
                    (parameterType != typeof(string)))
                {
                    continue;
                }

                method = _mi;
                break;
            }

            if (method == null)
            {
                throw new ConfigurationException(
                    "Configured method named '" + methodDesc.AccessorMethodName +
                    "' not found for class " + type.FullName);
            }

            return makeMethodDesc(method, methodDesc.Name);
        }
        
        /// <summary>
        /// Creates an event property descriptor for a field or native property
        /// </summary>
        /// <param name="type"></param>
        /// <param name="fieldDesc"></param>
        /// <returns></returns>

        private static EventPropertyDescriptor makeDesc(Type type, ConfigurationEventTypeLegacy.LegacyFieldPropDesc fieldDesc)
        {
            FieldInfo field = type.GetField( fieldDesc.AccessorFieldName ) ;
            if (field != null)
            {
                return makeFieldDesc(field, fieldDesc.Name);
            }

            PropertyDescriptorCollection properties = TypeDescriptor.GetProperties( type ) ;
            PropertyDescriptor property = properties[ fieldDesc.AccessorFieldName ] ;
            if ( property != null )
            {
                return makeNativeDesc(property, fieldDesc.Name);
            }

            throw new ConfigurationException(
                "Configured field named '" + fieldDesc.AccessorFieldName +
                "' not found for class " + type.FullName);
        }

        /// <summary>
        /// Makes a simple-type event property descriptor based on a reflected field.
        /// </summary>
        /// <param name="field">is the public field</param>
        /// <param name="name">is the name of the event property</param>
        /// <returns> property descriptor</returns>
        
        internal static EventPropertyDescriptor makeFieldDesc(FieldInfo field, String name)
        {
        	SimpleFieldPropertyDescriptor descriptor = new SimpleFieldPropertyDescriptor( field.Name, field ) ;
            return new EventPropertyDescriptor(name, name, descriptor, EventPropertyType.SIMPLE);
        }

        /// <summary>
        /// Makes a simple-type event property descriptor based on a reflected property.
        /// </summary>
        /// <param name="property">the property descriptor</param>
        /// <param name="name">the name of the event property</param>
        /// <returns> property descriptor</returns>

        internal static EventPropertyDescriptor makeNativeDesc(PropertyDescriptor property, String name)
        {
            return new EventPropertyDescriptor(name, name, property, EventPropertyType.SIMPLE);
        }

        /// <summary>
        /// Makes an event property descriptor based on a reflected method, considering
        /// the methods parameters to determine if this is an indexed or mapped event property.
        /// </summary>
        /// <param name="method">is the public method</param>
        /// <param name="name">is the name of the event property</param>
        /// <returns> property descriptor</returns>
        
        internal static EventPropertyDescriptor makeMethodDesc(MethodInfo method, String name)
        {
            EventPropertyType propertyType ;
            PropertyDescriptor descriptor = null ;

            ParameterInfo[] methodParameters = method.GetParameters();
            switch( methodParameters.Length ) {
            	case 0:
            		descriptor = new SimpleAccessorPropertyDescriptor( name, method ) ;
            		propertyType = EventPropertyType.SIMPLE;
            		break;
            	case 1:
	                Type parameterType = methodParameters[0].ParameterType;
	                descriptor = new IndexedAccessorPropertyDescriptor( name, method ) ;
	                if ( parameterType == typeof(string) ) {
		                propertyType = EventPropertyType.MAPPED ;
	                } else if ( parameterType == typeof(int) ) {
	                	propertyType = EventPropertyType.INDEXED ;
	                } else {
	                	throw new ArgumentException( "Method parameter was not of type string or int" ) ;
	                }
	                break ;
	            default:
	                throw new ArgumentException( "Method must have zero or one parameter" ) ;
            }
            
            return new EventPropertyDescriptor(name, name, descriptor, propertyType);
        }
    }
}
